cardano-node/src/Cardano/Node/Run.hs

```diff
@@ -263,3 +263,29 @@ handleSimpleNode p trace nodeTracers nCli nc = do
           then ValidateAllEpochs
           else ValidateMostRecentEpoch
       }
+
+exceptionHandler :: IO () -> IO ()
+exceptionHandler m = m `catchSync` \(e :: SomeException) -> do
+    let showOutput    = show e
+        displayOutput = displayException e
+    if showOutput /= displayOutput
+      then do
+        -- First print the detailed output of 'show' so that we at least log
+        -- the exact exception for debugging purposes.
+        putStrLn showOutput
+        -- Then print the human-readable output of 'displayException' last, so
+        -- that the last thing in the logs is understandable to the
+        -- user/sysadmin.
+        putStrLn displayOutput
+      else
+        -- If the human-readable output of 'displayException' is the same as
+        -- the output of 'show', just print it once
+        putStrLn showOutput
+
+    -- Stop the node
+    return ()
```

Comment:
I am glad you are only catching sync exceptions here, because some of the async exceptions should probably never be caught (even though we do exactly that in `network`).

However, I think returning `()` here is a bad idea. What happens if caller runs this in a `Control.Monad.forever` loop? You will get the exception, print it, `return ()` and the function will just be called again. I have a very similar function in the `explorer` which catches the exception, logs it and rethrows it (using `throwIO`). Is that not (at least from a semanitc POV) a better solution?

Reply:
A caller should not do this. The only caller so far is the `main` in the executable (indirectly via some `cardano-shell` code). But yeah, we don't know what will happen in the future, so I take your point.

I can change it so that the exception is rethrow(IO)n. That means that we shouldn't print the `showOutput` here as the top-level exception handler will already do this for us. This also means that we can't print the `displayOutput` *after* the `showOutput`, as intended. But I can live with that. After all, this is only intended to make life easier for users and sysadmins, not for developers.

Reply:
I have updated the PR to rethrow(IO).